File: pollardRho.js
/**
 * Pollard Rho Factorization Method
 *
 * @author Isaac Flaum
 */
const numberToFactor = 36287; // large integer we want to factor with pollard rho method

const seed = 2; // Integer seed value or guess

const numberOfIterations = 20; // number of iterations to run polynomial. Note this should be an even number to make
  //the guessing of (x_2n - x_n) reach the last n.

const polynomial = x => (x * x) + 1; // The polynomial f(x) = x^2 + 1

let polynomialIterations = []; // a list to store all of the recursive iterations of the polynomial

/**
 * Euclidian algorithm to find GCD of two integers. I changed the function a bit but og credit to Stack Overflow
 * https://stackoverflow.com/questions/17445231/js-how-to-find-the-greatest-common-divisor
 */
const gcd = (a, b) => {
  if ((typeof a !== 'number') || (typeof b !== 'number')) {
    throw new Error('One of the arguments was not a number!');
  }
  a = Math.abs(a);
  b = Math.abs(b);
  while(b) {
    var t = b;
    b = a % b;
    a = t;
  }
  return a;
};

// generate our guesses
let result = seed;
for (var i = 0; i < 20; i++) {
	result = polynomial(result) % numberToFactor;
	console.log(`Adding f(${i + 1}) = ${result} to the results list.`)
	polynomialIterations.push(result);
}

// test our guesses
for (var i = 0; i < numberOfIterations / 2; i++) {
	console.log(`Testing gcd of ${numberToFactor} and ${polynomialIterations[2*(i+ 1) - 1]} - ${polynomialIterations[i]}`)
    let d = gcd(polynomialIterations[2*(i+ 1) - 1] - polynomialIterations[i], numberToFactor);
    if (d > 1) {
    	console.log(`Found a factor: x = ${d}`);
    	process.exit(0);
    }
}

File: java/src/com/isaac/PollardRho.java
package com.isaac;

import java.math.BigInteger;
import java.util.LinkedList;
import java.util.List;

public class PollardRho {

    BigInteger integerToFactor; // the number to factor

    private BigInteger polynomial(BigInteger x) { return x.multiply(x).add(BigInteger.ONE); } // f(x) = x^2 + 1

    private List<BigInteger> xs; // list to store iterations of polynomial

    private BigInteger seed; // starting seed for polynomial

    private int numberOfIterations; // number of times to run algorithm

    // Constructor
    public PollardRho(BigInteger integerToFactor, BigInteger seed, int numberOfIterations) {
        this.integerToFactor = integerToFactor;
        this.seed = seed;
        this.numberOfIterations = numberOfIterations;
        xs = new LinkedList<BigInteger>();
    }

    private void generateXs() {
        BigInteger result = seed;
        for (int i = 0; i < numberOfIterations; i++) {
            result = polynomial(result).mod(integerToFactor);
            xs.add(result);
        }
    }

    private BigInteger crunchXs() {
        BigInteger d = BigInteger.ONE; // the gcd of integerToFactor and (x_2i - x_i)
        for (int i = 0; i < numberOfIterations / 2; i++) {
            d = integerToFactor.gcd(xs.get(2*(i+1) - 1).subtract(xs.get(i)));
            if (d.compareTo(BigInteger.ONE) >= 1) {
                return d;
            }
        }
        return BigInteger.ONE;
    }

    public void pollardRho() {
        generateXs();
        BigInteger result = crunchXs();
        System.out.print("Found a factor: " + result);
    }

}
